### DISCUSSION TITLE
Derivative query forms cartesian product

### DISCUSSION THREAD

**Discussion by j042** [tataratat/splinepy#248] [General]
currently, derivatives queries take 2 forms:

```python
# (1)
ders = spline.derivatives(q, [1, 1])

# (2)
ders = spline.derivatives(q, orders)
assert len(q) == len(orders)
```
I don't have any use case for `(2)` yet, but I often encounter the following use case

```python
ders = []
for q_order in q_orders:
    ders.append(spline.derivatives(q, q_order))
```
A suggestion would be,
```python
ders = spline.derivatives(q, q_orders)
assert len(q_orders) * len(q) == len(ders)
for i in range(len(q_orders):
    assert np.allclose(ders[i::len(q_orders)], spline.derivatives(q, q_orders[i]))
```
could be nice. what do you think @jzwar

**Comment by jzwar** [comment]
So a list of derivatives, all computed consecutively?

**Reply by j042** [reply]
I guess it can be an option?
```python
# option (1) - iterate queries first
queries = np.random.random((n_query, para_dim))
result1 = []
for q_order in np.eye(para_dim):
  result1.append(spline.derivative(queries, q_order))

# option (2) - iterate q_orders first
result2 = []
for q_order in np.eye(para_dim):
    for query in queries:
        result2.append(spline.derivative(query, q_order))
```

**Reply by j042** [reply]
we'd implement this in cpp

**Reply by jzwar** [reply]
For `np.eye` we have Jacobin already. I honestly do not really see another use-case. I don't like changing behaviour for slightly different inputs... This makes the code to unpredictable. Especially as this can become really expensive fast...

**Reply by j042** [reply]
It's not about `np.eye`. This makes a lot more sense than current multi-order query

**Reply by jzwar** [reply]
But it basically a cartesian product between a list of orders and queries, right? Do you want to drop the old layout?

**Reply by j042** [reply]
yes (hence, the title) and yes. I guess more direct question would be: do you use "old" layout anywhere?

**Reply by jzwar** [reply]
no, I don't. Never used it. Go for it

**Comment by j042** [comment]
@OberGue

**Comment by j042** [comment]
#337
